File: apps/modlink_plugin_scaffold/modlink_plugin_scaffold/core/spec.py
"""Shared data models for the scaffold wizard and generator."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

DriverKind = Literal["driver", "loop"]
PayloadType = Literal["signal", "raster", "field", "video"]
DataArrivalPattern = Literal["push", "poll", "unsure"]

DEVICE_ID_PATTERN = re.compile(r"^[a-z0-9_]+\.[0-9]{2,}$")
TOKEN_PATTERN = re.compile(r"[^a-z0-9_]+")


def sanitize_identifier(name: str) -> str:
    """Convert free-form input into a Python/package safe identifier."""
    value = re.sub(r"[^\w\s-]", "", str(name))
    value = re.sub(r"[-\s]+", "_", value)
    value = value.lower().strip("_")
    if value and value[0].isdigit():
        value = f"plugin_{value}"
    return value


def normalize_token(value: str) -> str:
    """Normalize a token used by device ids, modalities, or providers."""
    normalized = TOKEN_PATTERN.sub("_", str(value).strip().lower())
    normalized = re.sub(r"_+", "_", normalized).strip("._")
    return normalized


def to_pascal_case(name: str) -> str:
    parts = re.split(r"[_-]+", str(name))
    return "".join(word.capitalize() for word in parts if word)


def to_title_words(name: str) -> str:
    words = normalize_token(name).split("_")
    return " ".join(word.capitalize() for word in words if word)


def make_device_id(name: str, index: int = 1) -> str:
    base = normalize_token(name)
    if not base:
        raise ValueError("device base name must not be empty")
    ordinal = max(1, int(index))
    return f"{base}.{ordinal:02d}"


def normalize_device_id(device_id: str) -> str:
    normalized = str(device_id).strip().lower().replace("-", "_")
    normalized = re.sub(r"[^a-z0-9_.]+", "_", normalized)
    normalized = re.sub(r"_+", "_", normalized).strip("._")
    return normalized


def is_valid_device_id(device_id: str) -> bool:
    normalized = normalize_device_id(device_id)
    return bool(DEVICE_ID_PATTERN.fullmatch(normalized))


def unique_strings(values: list[str] | tuple[str, ...]) -> tuple[str, ...]:
    result: list[str] = []
    seen: set[str] = set()
    for raw in values:
        value = str(raw).strip()
        if not value or value in seen:
            continue
        result.append(value)
        seen.add(value)
    return tuple(result)


@dataclass(slots=True)
class StreamSpec:
    modality: str
    payload_type: PayloadType
    display_name: str
    sample_rate_hz: float
    chunk_size: int
    channel_names: tuple[str, ...] = ()
    unit: str | None = None
    raster_length: int | None = None
    field_height: int | None = None
    field_width: int | None = None
    video_height: int | None = None
    video_width: int | None = None

    def __post_init__(self) -> None:
        self.modality = normalize_token(self.modality)
        if not self.modality:
            raise ValueError("stream modality must not be empty")
        self.display_name = str(self.display_name).strip() or f"{to_title_words(self.modality)} Stream"
        self.sample_rate_hz = float(self.sample_rate_hz)
        if self.sample_rate_hz <= 0.0:
            raise ValueError("sample_rate_hz must be positive")
        self.chunk_size = int(self.chunk_size)
        if self.chunk_size <= 0:
            raise ValueError("chunk_size must be positive")
        self.channel_names = unique_strings(self.channel_names)

        if self.payload_type == "signal" and not self.channel_names:
            self.channel_names = ("ch1",)
        if self.payload_type == "raster" and not self.channel_names:
            self.channel_names = ("intensity",)
        if self.payload_type == "field" and not self.channel_names:
            self.channel_names = ("intensity",)
        if self.payload_type == "video" and not self.channel_names:
            self.channel_names = ("red", "green", "blue")

        if self.payload_type == "raster":
            self.raster_length = max(1, int(self.raster_length or 1))
        if self.payload_type == "field":
            self.field_height = max(1, int(self.field_height or 1))
            self.field_width = max(1, int(self.field_width or 1))
        if self.payload_type == "video":
            self.video_height = max(1, int(self.video_height or 1))
            self.video_width = max(1, int(self.video_width or 1))

    @property
    def channel_count(self) -> int:
        return len(self.channel_names)

    @property
    def metadata(self) -> dict[str, object]:
        payload: dict[str, object] = {}
        if self.unit:
            payload["unit"] = self.unit
        if self.payload_type == "raster" and self.raster_length is not None:
            payload["length"] = self.raster_length
        if self.payload_type == "field":
            payload["height"] = self.field_height
            payload["width"] = self.field_width
        if self.payload_type == "video":
            payload["height"] = self.video_height
            payload["width"] = self.video_width
        return payload

    @property
    def expected_shape(self) -> str:
        if self.payload_type == "signal":
            return f"[{self.channel_count}, {self.chunk_size}]"
        if self.payload_type == "raster":
            return f"[{self.channel_count}, {self.chunk_size}, {self.raster_length}]"
        if self.payload_type == "field":
            return f"[{self.channel_count}, {self.chunk_size}, {self.field_height}, {self.field_width}]"
        return f"[{self.channel_count}, {self.chunk_size}, {self.video_height}, {self.video_width}]"

    @property
    def loop_interval_ms_suggestion(self) -> int:
        return max(1, int(round(1000.0 * self.chunk_size / self.sample_rate_hz)))


@dataclass(slots=True)
class DriverSpec:
    plugin_name: str
    display_name: str
    device_id: str
    providers: tuple[str, ...]
    driver_kind: DriverKind
    driver_reason: str
    data_arrival: DataArrivalPattern
    streams: tuple[StreamSpec, ...]
    dependencies: tuple[str, ...] = ()
    project_name: str = field(init=False)
    class_name: str = field(init=False)
    entry_point_name: str = field(init=False)

    def __post_init__(self) -> None:
        plugin_identifier = sanitize_identifier(self.plugin_name)
        if not plugin_identifier:
            raise ValueError("plugin_name must contain at least one letter or number")
        self.plugin_name = plugin_identifier
        self.project_name = self.plugin_name.replace("_", "-")
        self.class_name = to_pascal_case(self.plugin_name)
        self.entry_point_name = self.plugin_name.replace("_", "")

        self.display_name = str(self.display_name).strip() or self.class_name
        self.device_id = normalize_device_id(self.device_id)
        if not is_valid_device_id(self.device_id):
            raise ValueError("device_id must match name.XX")
        normalized_providers = tuple(
            provider
            for provider in (normalize_token(item) for item in self.providers)
            if provider
        )
        self.providers = unique_strings(normalized_providers)
        if not self.providers:
            raise ValueError("at least one provider is required")
        self.driver_reason = str(self.driver_reason).strip()
        self.streams = tuple(self.streams)
        if not self.streams:
            raise ValueError("at least one stream is required")

        dependencies = list(self.dependencies)
        dependencies.extend(["modlink-sdk", "numpy>=2.3.3"])
        self.dependencies = unique_strings(dependencies)

    @property
    def driver_base_class(self) -> str:
        return "LoopDriver" if self.driver_kind == "loop" else "Driver"

    @property
    def providers_tuple(self) -> str:
        return repr(self.providers)

    @property
    def providers_display(self) -> str:
        return ", ".join(self.providers)

    @property
    def suggested_loop_interval_ms(self) -> int:
        return min(stream.loop_interval_ms_suggestion for stream in self.streams)


@dataclass(frozen=True, slots=True)
class ProjectContext:
    working_dir: Path


@dataclass(frozen=True, slots=True)
class ScaffoldPaths:
    project_dir: Path
    package_dir: Path
    pyproject_path: Path
    readme_path: Path
    init_path: Path
    factory_path: Path
    driver_path: Path

    def generated_files(self) -> tuple[Path, ...]:
        return (
            self.pyproject_path,
            self.readme_path,
            self.init_path,
            self.factory_path,
            self.driver_path,
        )

File: tools/modlink_plugin_scaffold/src/lib/spec.ts
import {z} from "zod";

import {dataArrivalOrder, driverKindOrder, getCopy, payloadOrder} from "./i18n.js";
import type {DataArrival, Draft, DriverKind, DriverSpec, Language, PayloadType, StreamDraft, StreamSpec, ValidationResult} from "./types.js";

const deviceIdPattern = /^[a-z0-9_]+\.[0-9]{2,}$/;
const stringSchema = z.string();

export function createDefaultStream(index: number): StreamDraft {
  const labelIndex = index + 1;
  return {
    modality: `stream_${labelIndex}`,
    displayName: `Stream ${labelIndex}`,
    payloadType: "signal",
    sampleRateHz: "250",
    chunkSize: "25",
    channelCount: "2",
    channelNames: "ch1, ch2",
    unit: "",
    rasterLength: "128",
    fieldHeight: "48",
    fieldWidth: "48",
    videoHeight: "480",
    videoWidth: "640",
  };
}

export function createDefaultDraft(): Draft {
  return {
    pluginName: "my-device",
    displayName: "",
    deviceId: "",
    providersText: "serial",
    dataArrival: "unsure",
    driverKind: "driver",
    dependenciesText: "",
    streams: [createDefaultStream(0)],
    selectedStreamIndex: 0,
  };
}

export function sanitizeIdentifier(value: string): string {
  const normalized = String(value)
    .replace(/[^\w\s-]/g, "")
    .replace(/[-\s]+/g, "_")
    .toLowerCase()
    .replace(/^_+|_+$/g, "");
  if (!normalized) {
    return "";
  }
  return /^\d/.test(normalized) ? `plugin_${normalized}` : normalized;
}

export function normalizeToken(value: string): string {
  return String(value)
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9_]+/g, "_")
    .replace(/_+/g, "_")
    .replace(/^_+|_+$/g, "");
}

export function toPascalCase(value: string): string {
  return String(value)
    .split(/[_-]+/g)
    .filter(Boolean)
    .map((part) => `${part[0]?.toUpperCase() ?? ""}${part.slice(1)}`)
    .join("");
}

export function toTitleWords(value: string): string {
  return normalizeToken(value)
    .split("_")
    .filter(Boolean)
    .map((part) => `${part[0]?.toUpperCase() ?? ""}${part.slice(1)}`)
    .join(" ");
}

export function makeDeviceId(pluginName: string): string {
  const normalized = normalizeToken(pluginName);
  return normalized ? `${normalized}.01` : "";
}

export function splitCsv(value: string): string[] {
  const items = String(value)
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean);
  return Array.from(new Set(items));
}

export function positiveInt(value: string): number | null {
  const parsed = Number.parseInt(String(value).trim(), 10);
  return Number.isInteger(parsed) && parsed > 0 ? parsed : null;
}

export function positiveFloat(value: string): number | null {
  const parsed = Number.parseFloat(String(value).trim());
  return Number.isFinite(parsed) && parsed > 0 ? parsed : null;
}

export function recommendedDriverKind(dataArrival: DataArrival): DriverKind {
  return dataArrival === "poll" ? "loop" : "driver";
}

export function driverReason(language: Language, dataArrival: DataArrival, driverKind: DriverKind): string {
  const copy = getCopy(language);
  if (dataArrival === "push") {
    return driverKind === "driver" ? copy.reasonPushDriver : copy.reasonPushLoop;
  }
  if (dataArrival === "poll") {
    return driverKind === "loop" ? copy.reasonPollLoop : copy.reasonPollDriver;
  }
  return driverKind === "driver" ? copy.reasonUnsureDriver : copy.reasonUnsureLoop;
}

export function applyPluginName(draft: Draft, value: string): Draft {
  const previousPluginName = draft.pluginName;
  const previousDefaultDisplay = toPascalCase(sanitizeIdentifier(previousPluginName) || previousPluginName);
  const previousDefaultDeviceId = makeDeviceId(previousPluginName);
  const normalized = sanitizeIdentifier(value);

  return {
    ...draft,
    pluginName: value,
    displayName:
      !draft.displayName.trim() || draft.displayName === previousDefaultDisplay
        ? (normalized ? toPascalCase(normalized) : "")
        : draft.displayName,
    deviceId:
      !draft.deviceId.trim() || draft.deviceId === previousDefaultDeviceId
        ? makeDeviceId(value)
        : draft.deviceId,
  };
}

export function applyDataArrival(draft: Draft, value: DataArrival): Draft {
  if (!dataArrivalOrder.includes(value)) {
    return draft;
  }
  const previousRecommended = recommendedDriverKind(draft.dataArrival);
  return {
    ...draft,
    dataArrival: value,
    driverKind: draft.driverKind === previousRecommended ? recommendedDriverKind(value) : draft.driverKind,
  };
}

export function applyPayloadDefaults(stream: StreamDraft, payloadType: PayloadType): StreamDraft {
  if (payloadType === "signal") {
    return {
      ...stream,
      payloadType,
      sampleRateHz: "250",
      chunkSize: "25",
      channelCount: "2",
      channelNames: "ch1, ch2",
    };
  }
  if (payloadType === "raster") {
    return {
      ...stream,
      payloadType,
      sampleRateHz: "10",
      chunkSize: "1",
      channelCount: "1",
      channelNames: "intensity",
      rasterLength: "128",
    };
  }
  if (payloadType === "field") {
    return {
      ...stream,
      payloadType,
      sampleRateHz: "10",
      chunkSize: "1",
      channelCount: "1",
      channelNames: "intensity",
      fieldHeight: "48",
      fieldWidth: "48",
    };
  }
  return {
    ...stream,
    payloadType,
    sampleRateHz: "30",
    chunkSize: "1",
    channelCount: "3",
    channelNames: "red, green, blue",
    videoHeight: "480",
    videoWidth: "640",
    unit: "",
  };
}

function validateStream(language: Language, stream: StreamDraft, index: number, fieldErrors: Record<string, string>): StreamSpec | null {
  const copy = getCopy(language);
  const prefix = `streams.${index}`;
  const modality = normalizeToken(stream.modality);

  if (!modality) {
    fieldErrors[`${prefix}.modality`] = copy.modalityError;
  }

  if (!payloadOrder.includes(stream.payloadType)) {
    fieldErrors[`${prefix}.payloadType`] = copy.validationBlocked;
    return null;
  }

  const sampleRateHz = positiveFloat(stream.sampleRateHz);
  if (sampleRateHz === null) {
    fieldErrors[`${prefix}.sampleRateHz`] = copy.positiveFloatError;
  }

  const chunkSize = positiveInt(stream.chunkSize);
  if (chunkSize === null) {
    fieldErrors[`${prefix}.chunkSize`] = copy.positiveIntError;
  }

  const channelNames = splitCsv(stream.channelNames);
  if (channelNames.length === 0) {
    fieldErrors[`${prefix}.channelNames`] = copy.channelNamesError;
  }

  const channelCount = positiveInt(stream.channelCount);
  if (stream.payloadType === "signal") {
    if (channelCount === null) {
      fieldErrors[`${prefix}.channelCount`] = copy.positiveIntError;
    } else if (channelNames.length > 0 && channelNames.length !== channelCount) {
      fieldErrors[`${prefix}.channelNames`] = copy.signalChannelNamesCountError;
    }
  }

  const rasterLength = stream.payloadType === "raster" ? positiveInt(stream.rasterLength) : null;
  if (stream.payloadType === "raster" && rasterLength === null) {
    fieldErrors[`${prefix}.rasterLength`] = copy.positiveIntError;
  }

  const fieldHeight = stream.payloadType === "field" ? positiveInt(stream.fieldHeight) : null;
  const fieldWidth = stream.payloadType === "field" ? positiveInt(stream.fieldWidth) : null;
  if (stream.payloadType === "field") {
    if (fieldHeight === null) {
      fieldErrors[`${prefix}.fieldHeight`] = copy.positiveIntError;
    }
    if (fieldWidth === null) {
      fieldErrors[`${prefix}.fieldWidth`] = copy.positiveIntError;
    }
  }

  const videoHeight = stream.payloadType === "video" ? positiveInt(stream.videoHeight) : null;
  const videoWidth = stream.payloadType === "video" ? positiveInt(stream.videoWidth) : null;
  if (stream.payloadType === "video") {
    if (videoHeight === null) {
      fieldErrors[`${prefix}.videoHeight`] = copy.positiveIntError;
    }
    if (videoWidth === null) {
      fieldErrors[`${prefix}.videoWidth`] = copy.positiveIntError;
    }
  }

  if (Object.keys(fieldErrors).some((key) => key.startsWith(`${prefix}.`))) {
    return null;
  }

  return {
    modality,
    displayName: stringSchema.parse(stream.displayName).trim() || `${toTitleWords(modality)} ${copy.defaultStreamName}`,
    payloadType: stream.payloadType,
    sampleRateHz: sampleRateHz ?? 1,
    chunkSize: chunkSize ?? 1,
    channelNames,
    unit: stream.unit.trim() || undefined,
    rasterLength: rasterLength ?? undefined,
    fieldHeight: fieldHeight ?? undefined,
    fieldWidth: fieldWidth ?? undefined,
    videoHeight: videoHeight ?? undefined,
    videoWidth: videoWidth ?? undefined,
  };
}

export function validateDraft(language: Language, draft: Draft): ValidationResult {
  const copy = getCopy(language);
  const fieldErrors: Record<string, string> = {};

  const pluginName = sanitizeIdentifier(draft.pluginName);
  if (!pluginName) {
    fieldErrors.pluginName = copy.pluginNameError;
  }

  const deviceId = (draft.deviceId.trim() || makeDeviceId(draft.pluginName)).trim().toLowerCase().replace(/-/g, "_");
  if (!deviceIdPattern.test(deviceId)) {
    fieldErrors.deviceId = copy.deviceIdError;
  }

  const providers = splitCsv(draft.providersText).map(normalizeToken).filter(Boolean);
  if (providers.length === 0) {
    fieldErrors.providersText = copy.providersError;
  }

  if (draft.streams.length === 0) {
    fieldErrors.streams = copy.streamsRequiredError;
  }

  const streamSpecs = draft.streams
    .map((stream, index) => validateStream(language, stream, index, fieldErrors))
    .filter((stream): stream is StreamSpec => stream !== null);

  let spec: DriverSpec | null = null;
  if (Object.keys(fieldErrors).length === 0) {
    spec = {
      pluginName,
      projectName: pluginName.replace(/_/g, "-"),
      className: toPascalCase(pluginName),
      displayName: draft.displayName.trim() || toPascalCase(pluginName),
      deviceId,
      providers,
      driverKind: driverKindOrder.includes(draft.driverKind) ? draft.driverKind : "driver",
      driverReason: driverReason(language, draft.dataArrival, draft.driverKind),
      dataArrival: draft.dataArrival,
      dependencies: Array.from(new Set(["modlink-sdk", "numpy>=2.3.3", ...splitCsv(draft.dependenciesText)])),
      streams: streamSpecs,
    };
  }

  return {
    spec,
    fieldErrors,
    recommendedDriverKind: recommendedDriverKind(draft.dataArrival),
    recommendedReason: driverReason(language, draft.dataArrival, recommendedDriverKind(draft.dataArrival)),
  };
}

export function visibleStreamFieldKeys(payloadType: PayloadType): string[] {
  if (payloadType === "signal") {
    return ["channelCount", "channelNames", "unit"];
  }
  if (payloadType === "raster") {
    return ["channelNames", "unit", "rasterLength"];
  }
  if (payloadType === "field") {
    return ["channelNames", "unit", "fieldHeight", "fieldWidth"];
  }
  return ["channelNames", "videoHeight", "videoWidth"];
}
